### PR TITLE
Implementation of ring gossiping based on memberlist

### PR DESCRIPTION
**Pull request by pstibrany**
See issue #1721 for details about the implementation.

Metrics are now included in the implementation. Upgrade path will be covered by separate issue, #1525.

### CODE REVIEW

File: pkg/ring/kv/memberlist/broadcast.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+type ringBroadcast struct {
+	logger   log.Logger
+	key      string
+	content  []string // Description of what is stored in this value. Used for invalidation.
+	version  uint     // local version of the value, generated by merging this change
+	msg      []byte   // encoded key and value
+	finished func(b ringBroadcast)
+}
+
```
Comment (gouthamve) [inline pkg/ring/kv/memberlist/broadcast.go]: Maybe add a comment that it implements `memberlist.Broadcast`. Was wondering where this was being called.

File: pkg/ring/kv/memberlist/memberlist_client.go [hunk 2]
```diff
@@ -0,0 +1,853 @@
...
+	if cfg.NodeName != "" {
+		mlCfg.Name = cfg.NodeName
+	}
+
+	mlCfg.LogOutput = newMemberlistLoggerAdapter(logger, false)
+	mlCfg.Transport = tr
+
+	// Memberlist uses UDPBufferSize to figure out how many messages it can put into single packet.
+	// As we don't use UDP for sending packets, we can use higher value here.
+	if mlCfg.UDPBufferSize < 16384 {
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ring/kv/memberlist/memberlist_client.go]: I'm not sure what I was thinking here. Default (1400) is too small, and I guess I wanted to leave some room for memberlist default going up. It's a silly condition. If you ask the same question for value I write here instead (10MB), I just wanted to express that going to MBs is fine here, as we use TCP. In practice, you can only get to such high value if there are many queued broadcast messages waiting, as our messages are still pretty small (tens of KBs).

File: pkg/ring/kv/memberlist/timestamps.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
+package memberlist
+
+import (
+	"bytes"
+	"encoding/gob"
+	"fmt"
+	"time"
+)
+
+// This is a simple state that is gossiped under specific key by memberlist_client itself.
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ring/kv/memberlist/timestamps.go]: Again, monitoring of gossip performance. Trying to answer this question: how do we know if this is broken? One possibility is that everything is OK if "now - last_heard_from < ~30s", as measured for every node pair. But that's too many combinations as you point out.

File: pkg/ring/kv/memberlist/memberlist_client.go [hunk 4]
```diff
@@ -0,0 +1,853 @@
...
+	if len(key) > maxKeyLength {
+		return fmt.Errorf("key too long: %d", len(key))
+	}
+
+	sleep := false
+	for retries := 10; retries > 0; retries-- {
+		m.casAttempts.Inc()
+
+		if sleep {
+			// wait for a while before retrying... typically merge fails because timestamp hasn't
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ring/kv/memberlist/memberlist_client.go]: This is a workaround for a problem in Ring merge function. Merging is based on timestamps with 1 second resolution. If we call 'f' function several times in a row with no sleep, it will simply report the same timestamp, and then merge will not do any change. Solution is either to make merge more complex (which I wanted to avoid), or simply sleep for a short time in a hope that next call of 'f' will put new timestamp. (Another solution would be to introduce new Timestamp to the ring, eg. with nanosecond resolution)

So basically this is a hack to help Ring merging.

File: pkg/ring/kv/memberlist/broadcast.go (hunk 1 above)
Comment (pstibrany) [reply pkg/ring/kv/memberlist/broadcast.go]: That's a good idea, will do.

File: pkg/ring/kv/memberlist/metrics.go [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+
+	ch <- prometheus.MustNewConstMetric(m.storeValuesDesc, prometheus.GaugeValue, float64(len(m.store)))
+
+	for k, v := range m.store {
+		ch <- prometheus.MustNewConstMetric(m.storeSizesDesc, prometheus.GaugeValue, float64(len(v.value)), k)
+	}
+}
+
+// Runs a goroutine that updates timestamp for this node every couple of seconds
+func (m *Client) updateNodeTimestamp() {
```
[the remark this answers is not in the document]
Comment (gouthamve) [reply pkg/ring/kv/memberlist/metrics.go]: I think we can assume that each node is gossiping every 15s (it's node update). And we will have a HTML page that shows how old each gossip was.

File: pkg/ring/kv/memberlist/timestamps.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (gouthamve) [reply pkg/ring/kv/memberlist/timestamps.go]: See, above. We are going to gossip every 5/15s anyways for the ring. We could use the same thing.

File: pkg/ring/kv/memberlist/memberlist_client.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (gouthamve) [reply pkg/ring/kv/memberlist/memberlist_client.go]: Why are we failing if there is no merge? I would assume that if there is no change, then we will be succeeding. The new value we want to set is already set....

[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ring/kv/memberlist/memberlist_client.go]: There is a mismatch between two functions, 'f' and Merge.

It is up to function 'f' to detect no change, and return nil in such case. If 'f' returns new value, it is trying to write it to the store.

If Merge returns no change, it only means that Merge didn't *detect* the change, but since 'f' has reported new value, we can expect that there is *some* change. By retrying several times, the hope is that Merge will be able to see it.

(Sleep part is there because we know that Ring's Merge depends on timestamps with second-resolution.)

You may also ask, why do we use Merge in CAS in the first place? There are two reasons for that: 1) it allows us to find out what the change was and broadcast only this change, 2) merging is what other gossiping members do when they receive our state. If we don't detect any change, it's most likely that they won't either, and will ignore such message.

File: pkg/ring/kv/memberlist/metrics.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ring/kv/memberlist/metrics.go]: "it's a node update" -- you mean ingesters, right? I wanted to cover more than just ingesters here... all components that read the ring.

When you're talking about HTML page, do you mean existing /ring page? Or something new?

By exporting the metrics, I wanted to have an ability to setup alerts if things go too wrong.

(I'm not saying this is good solution – it's not, just explaining the thinking behind it.)

File: pkg/ring/kv/memberlist/timestamps.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ring/kv/memberlist/timestamps.go]: I will resolve this to keep discussion in one thread only.

File: pkg/ring/kv/memberlist/metrics.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (gouthamve) [reply pkg/ring/kv/memberlist/metrics.go]: > By exporting the metrics, I wanted to have an ability to setup alerts if things go too wrong.

Could you list the alerts (what you want to alert for) here? Maybe we can come up with alternative ways to achieve that?

[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ring/kv/memberlist/metrics.go]: I remember discussing this with you, but not if we have found a solution.

Alert we were thinking about is roughly this ... make sure that "now - last_heard_from < ~30s", where "last heard from" is every other node.

In practice, I think we need to make sure that all ingesters are up-to-date, and since we already have ingesters and their latest timestamp in the ring, we can use that. So what about creating a two metrics:

1) number of non-left ingesters in the ring
2) oldest timestamp of non-left ingester in the ring

If 1) is greater than 0, and 2) is greater than (say) 30 seconds, we can raise an alert. These two metrics would be exposed by all components that use the ring.

How does that sound?

Now this solution is fine (hopefully) for ring with ingesters, but if we reuse memberlist KV store for something else, we need new metrics for that.

File: pkg/ring/kv/memberlist/memberlist_client.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ring/kv/memberlist/memberlist_client.go]: Does this make sense now? Can I resolve it?

File: pkg/ring/kv/memberlist/memberlist_client.go [hunk 6]
```diff
@@ -0,0 +1,851 @@
...
+
+	// Coded used to encode/decode the value
+	codec codec.Codec
+}
+
+// NewMemberlistClient creates new Client instance. If cfg.JoinMembers is set, it will also try to connect
+// to these members and join the cluster. If that fails and AbortIfJoinFails is true, error is returned and no
+// client is created.
+func NewMemberlistClient(cfg Config, codec codec.Codec) (*Client, error) {
+	logger := cfg.Logger
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ring/kv/memberlist/memberlist_client.go]: I've developed memberlist client outside of Cortex first, this is a leftover from that.

[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ring/kv/memberlist/memberlist_client.go]: Personally, I think it's a good practice to inject dependencies this way, instead of reaching to global logger value, but I'll change the code to fit existing Cortex codebase, and not introduce new patterns. Sorry for that.

File: pkg/ring/kv/memberlist/memberlist_client.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (gouthamve) [reply pkg/ring/kv/memberlist/memberlist_client.go]: Yes please. Just mentioned that changes in merge depend on the timestamp and not the tokens?

File: pkg/ring/kv/memberlist/memberlist_client.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (gouthamve) [reply pkg/ring/kv/memberlist/memberlist_client.go]: Haha, I agree with you, no need to be sorry. But we are stuck with this pattern and it's best to be consistent everywhere.

File: pkg/ring/kv/memberlist/memberlist_client.go [hunk 7]
```diff
@@ -0,0 +1,851 @@
...
+	if logger == nil {
+		logger = util.Logger
+	}
+
+	transportLogger := logger
+	if !cfg.TransportDebug {
+		transportLogger = level.NewFilter(transportLogger, level.AllowInfo())
+	}
+	tr, err := NewTCPTransport(&TCPTransportConfig{
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ring/kv/memberlist/memberlist_client.go]: Done.

File: pkg/ring/kv/memberlist/memberlist_client.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ring/kv/memberlist/memberlist_client.go]: Done.

File: pkg/ingester/transfer.go [hunk 8]
```diff
@@ -81,6 +81,33 @@ func (i *Ingester) TransferChunks(stream client.Ingester_TransferChunksServer) e
 		if fromIngesterID == "" {
 			fromIngesterID = wireSeries.FromIngesterId
 			level.Info(util.Logger).Log("msg", "processing TransferChunks request", "from_ingester", fromIngesterID)
+
+			// Ring gossiping: check if "from" ingester is in LEAVING state. It should be, but we may not see that yet.
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ingester/transfer.go]: Done.

File: pkg/ring/kv/memberlist/metrics.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ring/kv/memberlist/metrics.go]: I've removed timestamps gossiping, and per-value codec. Need to add those two new metrics now.

File: pkg/ring/replication_strategy.go [hunk 9]
```diff
@@ -47,6 +47,12 @@ func (r *Ring) replicationStrategy(ingesters []IngesterDesc, op Operation) ([]In
 
 // IsHealthy checks whether an ingester appears to be alive and heartbeating
 func (r *Ring) IsHealthy(ingester *IngesterDesc, op Operation) bool {
+	if ingester.State == LEFT {
+		if op == Reporting {
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ring/replication_strategy.go]: op == Reporting is only used in two places
1) on /ring page
2) when sending metrics

In both cases, I prefer to see these ingesters as "LEFT" and not "Unhealthy", which would be the case if IsHealthy reported 'false' here. They are not really unhealthy – they are gone :)

For other purposes (reading/writing), we report unhealthy for these, so they are ignored.

Perhaps I should put this comment in a code instead.

File: pkg/ring/lifecycler.go [hunk 10]
```diff
@@ -321,10 +328,39 @@ loop:
...
 					os.Exit(1)
 				}
+
+				if targetState == JOINING {
+					level.Info(util.Logger).Log("msg", "observing tokens before going ACTIVE")
+					observeChan = time.After(i.cfg.ObservePeriod)
+				}
+			}
+
+		case <-observeChan:
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ring/lifecycler.go]: If observeChan is nil, this case is never hit. By setting it to nil or non-nil values, code is effectively disabling (nil)/enabling (non-nil) this case.

File: pkg/ring/model.go [hunk 11]
```diff
@@ -73,48 +75,105 @@ func (d *Desc) RemoveIngester(id string) {
...
+	d.Tokens = output
+}
+
 // ClaimTokens transfers all the tokens from one ingester to another,
 // returning the claimed token.
 func (d *Desc) ClaimTokens(from, to string, normaliseTokens bool) []uint32 {
 	var result []uint32
 
-	if normaliseTokens {
+	// If the ingester we are claiming from is normalising, get its tokens then erase them from the ring.
```
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ring/model.go]: Unfortunately, the old code didn't cover all the cases (eg. when not doing normalization, but claiming tokens from ingester which is doing normalization -- quite common occurrence with normalizing merge function, or possibly when doing rollback of failed upgrade to normalized ingesters).

I've tried to adapt old code to missing edge cases, but then gave up and wrote it from scratch to make sure I cover all cases I thought are missing (and also added unit tests).

I agree that it is quite complex, and I can give it one more try to simplify it.

File: pkg/ring/lifecycler.go [hunk 12]
```diff
@@ -401,12 +439,78 @@ func (i *Lifecycler) initRing(ctx context.Context) error {
...
 		level.Info(util.Logger).Log("msg", "existing entry found in ring", "state", i.GetState(), "tokens", len(tokens))
-		return ringDesc, true, nil
+		// we haven't modified the ring, don't try to store it.
+		return nil, true, nil
 	})
 }
 
+// Verifies that tokens that this ingester has registered to the ring still belong to it.
+// Gossiping ring may change the ownership of tokens in case of conflicts.
+// If ingester doesn't its tokens anymore, this method generates new ones and puts them to the ring.
```
Comment (pstibrany) [inline pkg/ring/lifecycler.go]: doesn't own all

File: pkg/ring/replication_strategy.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (gouthamve) [reply pkg/ring/replication_strategy.go]: Yep a comment would be good.

**Comment by pstibrany** [comment]
> Left a few more comments now, sorry 🙃
:-) I really appreciate you taking time to review this. Thanks!

> Also we should mark every thing as experimental. We haven't run this yet and we don't know how it will perform in production, and I don't want users to use it yet. Could you build something like this https://github.com/cortexproject/cortex/blob/master/pkg/util/flagext/deprecated.go to indicate the flags are experimental?

I don't think I can do it in the same way. deprecatedFlag only logs message, but doesn't set any value. ExperimentalFlag would need to wrap existing flag.Value, which would be fine, if existing Value types in flag package were either exported (they are not), or if there was a way to replace a flag in FlagSet (I cannot find it). Or perhaps I can use Value types from other packages. But it feels unnecessary... what about I simply log a message when someone tries to use memberlist_client?

> Could you add a changelog entry too?

Sure.

**Comment by gouthamve** [comment]
> what about I simply log a message when someone tries to use memberlist_client?

That should be good yep. Also, mention EXPERIMENTAL in the `mlclient.join` description?

And rename the flag from `mlclient` to `memberlist`?

File: pkg/ring/lifecycler.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (gouthamve) [reply pkg/ring/lifecycler.go]: Thanks, TIL!

File: pkg/ring/model.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (pstibrany) [reply pkg/ring/model.go]: I've reverted to previous version of ClaimTokens, and only added missing case of non-normalising ingester claiming tokens from normalising (eg. in rollback scenario).

File: pkg/ring/kv/memberlist/memberlist_client.go [hunk 13]
```diff
@@ -467,7 +475,8 @@ func (m *Client) CAS(ctx context.Context, key string, f func(in interface{}) (ou
 				break
 			}
 
-			sleep = true
+			// If merge failed because of version mismatch, don't sleep.
+			sleep = err != errVersionMismatch
```
Comment (gouthamve) [inline pkg/ring/kv/memberlist/memberlist_client.go]: Could we have `sleep = (err != errVersionMismatch)`? Better readability

Comment (pstibrany) [reply pkg/ring/kv/memberlist/memberlist_client.go]: 👍

File: pkg/ring/kv/memberlist/metrics.go [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+		Help:      "Number of dropped notifications in WatchPrefix function",
+	}, []string{"prefix"})
+
+	if m.cfg.MetricsRegisterer == nil {
+		return
+	}
+
+	// if registration fails, that's too bad, but don't panic
+	for _, c := range m.getAllMetrics() {
+		err := m.cfg.MetricsRegisterer.Register(c.(prometheus.Collector))
```
Comment (jtlisi) [inline pkg/ring/kv/memberlist/metrics.go]: Why not panic? It shouldn't be possible to pass the same register to different MemberListClients and if it does happen it should happen relatively quickly, correct?

File: pkg/ring/kv/memberlist/metrics.go [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+		}
+	}
+
+	err := m.cfg.MetricsRegisterer.Register(m)
+	if err != nil {
+		level.Error(util.Logger).Log("msg", "failed to register metrics", "err", err)
+	}
+}
+
+func (m *Client) unregisterMetrics() {
```
Comment (jtlisi) [inline pkg/ring/kv/memberlist/metrics.go]: Is there a reason we need this kind of functionality? Seems a bit heavy for a client that we are only going to instantiate once?

File: pkg/ring/kv/memberlist/memberlist_client.go [hunk 16]
```diff
@@ -0,0 +1,788 @@
...
+	JoinMembers      flagext.StringSlice `yaml:"join_members"`
+	AbortIfJoinFails bool                `yaml:"abort_if_cluster_join_fails"`
+
+	// Remove LEFT ingesters from ring after this timeout.
+	LeftIngestersTimeout time.Duration `yaml:"left_ingesters_timeout"`
+
+	// Timeout used when leaving the memberlist cluster.
+	LeaveTimeout time.Duration `yaml:"leave_timeout"`
+
+	TCPTransport TCPTransportConfig
```
Comment (jtlisi) [inline pkg/ring/kv/memberlist/memberlist_client.go]: Can you add a yaml struct tag here?

File: docs/arguments.md [hunk 17]
```diff
@@ -142,6 +142,39 @@ prefix these flags with `distributor.ha-tracker.`
 - `etcd.max-retries`
    The maximum number of retries to do for failed ops.
 
+#### memberlist
+
+Flags for configuring KV store based on memberlist library.
```
Comment (khaines) [inline docs/arguments.md]: Should there be a callout at the start of this section regarding this feature (& its flags) being experimental?

File: pkg/ring/kv/memberlist/memberlist_logger.go [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+)
+
+// StdlibAdapter wraps a Logger and allows it to be passed to the stdlib
+// logger's SetOutput. It will extract date/timestamps, filenames, and
+// messages, and place them under relevant keys.
+type loggerAdapter struct {
```
Comment (jtlisi) [inline pkg/ring/kv/memberlist/memberlist_logger.go]: This is a nifty little struct. It seems more tied to the stdlib and not memberlist. Maybe move this to the utils package.

File: pkg/ring/kv/memberlist/memberlist_client.go [hunk 19]
```diff
@@ -0,0 +1,788 @@
...
+					level.Warn(util.Logger).Log("msg", "failed to send notification to prefix watcher", "prefix", p)
+				}
+			}
+		}
+	}
+}
+
+// CAS is part of kv.Client implementation.
+//
+// Unlike other implementation, it doesn't really do Compare-And-Swap. New values returned from function 'f'
```
Comment (pstibrany) [inline pkg/ring/kv/memberlist/memberlist_client.go]: Comment is slightly outdated... this function combines CAS and Merge together. Value is CAS-ed only if Merge returns a change.

File: pkg/ring/kv/memberlist/memberlist_logger.go [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+import (
+	"io"
+	"regexp"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+)
+
+// StdlibAdapter wraps a Logger and allows it to be passed to the stdlib
```
Comment (pstibrany) [inline pkg/ring/kv/memberlist/memberlist_logger.go]: This comment is a leftover from original. Should be fixed.

File: docs/arguments.md (hunk 17 above)
Comment (pstibrany) [reply docs/arguments.md]: Added EXPERIMENTAL warning here.

File: pkg/ring/kv/memberlist/memberlist_logger.go (hunk 18 above)
Comment (pstibrany) [reply pkg/ring/kv/memberlist/memberlist_logger.go]: This is in fact copy of https://godoc.org/github.com/go-kit/kit/log#StdlibAdapter, with modifications to understand output produced by memberlist library (esp. log level). See usage of m.logger here: https://github.com/hashicorp/memberlist/blob/master/memberlist.go

I've updated the comment to reflect that.

File: pkg/ring/kv/memberlist/metrics.go (hunk 14 above)
Comment (pstibrany) [reply pkg/ring/kv/memberlist/metrics.go]: You're right. This came from failing unit tests that tried to register the same metrics, but then I added MetricsRegisterer to the config, which fixed the problem as well. This isn't necessary anymore, same as unregister.

File: pkg/ring/kv/memberlist/metrics.go (hunk 15 above)
Comment (pstibrany) [reply pkg/ring/kv/memberlist/metrics.go]: It's gone now. See above comment.

**Comment by pstibrany** [comment]
Rebased on master, and added LEAVING state verification in new TransferTSDB method.

**Comment by pstibrany** [comment]
I've rebased onto master, and squashed all commits into one, so that future rebasing is easier ;-)

File: pkg/ring/kv/memberlist/memberlist_logger.go (hunk 18 above)
Comment (jtlisi) [reply pkg/ring/kv/memberlist/memberlist_logger.go]: Cool
